src/proxy_protocol/mod.rs

```diff
@@ -0,0 +1,571 @@
+        return Err(io::Error::new(e.kind(), msg));
+    }
+
+    // 2. check for v2 prefix
+    if &buffer[..V2_PREFIX_LEN] != v2::PROTOCOL_PREFIX {
+        if &buffer[..V1_PREFIX_LEN] == v1::PROTOCOL_PREFIX.as_bytes() {
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidData,
+                "V1 Proxy Protocol header detected, which is not supported",
+            ));
+        } else {
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidData,
+                "No valid Proxy Protocol header detected",
+            ));
+        }
+    }
+
+    // 3. decode the v2 header length
+    let length = u16::from_be_bytes([buffer[LENGTH], buffer[LENGTH + 1]]) as usize;
+    let full_length = V2_MINIMUM_LEN + length;
+
+    if full_length > DEFAULT_BUFFER_LEN {
+        return Err(io::Error::new(
+            io::ErrorKind::InvalidData,
+            format!("V2 Proxy Protocol header length is too long. Consider increasing default buffer size. Header length: {}", full_length),
+        ));
+    }
+
+    // 4. read the remaining header length
+    if let Err(e) = stream
+        .read_exact(&mut buffer[V2_MINIMUM_LEN..full_length])
+        .await
+    {
+        let msg = format!("Stream `read_exact` error: {}. Not able to read enough bytes to read the full V2 Proxy Protocol header.", e);
+        return Err(io::Error::new(e.kind(), msg));
+    }
+
+    // 5. parse the header
+    let header = HeaderResult::parse(&buffer[..full_length]);
+
+    match header {
+        HeaderResult::V1(Ok(_header)) => Err(io::Error::new(
+            io::ErrorKind::InvalidData,
+            "V1 Proxy Protocol header detected when parsing data",
+        )),
+        HeaderResult::V2(Ok(header)) => {
+            let client_address = match header.addresses {
+                v2::Addresses::IPv4(ip) => {
+                    SocketAddr::new(IpAddr::V4(ip.source_address), ip.source_port)
+                }
+                v2::Addresses::IPv6(ip) => {
+                    SocketAddr::new(IpAddr::V6(ip.source_address), ip.source_port)
+                }
+                v2::Addresses::Unix(unix) => {
+                    return Err(io::Error::new(
+                        io::ErrorKind::InvalidData,
+                        format!(
+                            "Unix socket addresses are not supported. Addresses: {:?}",
+                            unix
+                        ),
+                    ));
+                }
+                v2::Addresses::Unspecified => {
+                    // V2 PROXY header addresses "Unspecified"
+                    // Return client address as `None` so that "unknown" is used in the http header
+                    return Ok((stream, None));
+                }
+            };
+
+            Ok((stream, Some(client_address)))
+        }
+        HeaderResult::V1(Err(_error)) => Err(io::Error::new(
```

Comment:
This should be a panic once you detect v1 above if this is hit because is unsupported and should never hit.